src/HashingLogic.ts

```diff
@@ -53,35 +120,198 @@ export const orderedStringify = (obj: {}) => {
+  const linkHash = hashMessage(orderedStringify(attestation.link))
+  const auxHash = hashMessage(attestation.aux)
+  return getMerkleTreeFromLeaves([dataHash, typeHash, linkHash, auxHash])
+}
+
+/**
+ * Given the contents of an attestation node, return the root hash of the Merkle tree
+ */
+export const hashAttestationNode = (attestation: IAttestationNode): Buffer => {
+  const dataTree = getDataTree(attestation)
+  return dataTree.getRoot()
+}
+
+/**
+ * Sign a buffer with a given private key and return a hex string of the signature
+ * @param hash Any message buffer
+ * @param privKey A private key buffer
+ */
+export const signHash = (hash: Buffer, privKey: Buffer): string => {
+  const sig = ethUtil.ecsign(hash, privKey)
+  return ethSigUtil.concatSig(sig.v, sig.r, sig.s)
+}
+
+/**
+ * Recover the address of the signer of a given message hash
+ * @param hash Buffer of the message that was signed
+ * @param sig Hex string of the signature
+ */
+export const recoverHashSigner = (hash: Buffer, sig: string) => {
+  const signature = ethUtil.toBuffer(sig)
+  const sigParams = ethUtil.fromRpcSig(signature)
+  const pubKey = ethUtil.ecrecover(hash, sigParams.v, sigParams.r, sigParams.s)
+  const sender = ethUtil.publicToAddress(pubKey)
+  return ethUtil.bufferToHex(sender)
+}
+
+/**
+ * Given an array of hashed dataNode signatures and a hashed checksum signature, creates a new MerkleTree
+ * after padding, and sorting.
+ *
+ */
+export const getBloomMerkleTree = (
+  dataHashes: string[],
+  paddingNodes: string[],
+  checksumHash: string
+): MerkleTree => {
+  let leaves = dataHashes
+  leaves.push(checksumHash)
+  leaves = leaves.concat(paddingNodes)
   return getMerkleTreeFromLeaves(leaves)
 }
 
+/**
+ * Given an array of root hashes, sort and hash them into a checksum buffer
+ * @param {string[]} dataHashes - array of dataHashes as hex strings
+ */
+export const getChecksum = (dataHashes: string[]): Buffer => {
+  return ethUtil.toBuffer(hashMessage(JSON.stringify(dataHashes.sort())))
+}
+
+/**
+ * Given an array of root hashes, get and sign the checksum
+ * @param dataHashes - array of dataHashes as hex strings
+ * @param privKey - private key of signer
+ */
+export const signChecksum = (dataHashes: string[], privKey: Buffer): string => {
+  return signHash(getChecksum(dataHashes), privKey)
+}
+
+/**
+ * Sign a complete attestation node and return an object containing the datanode and the signature
+ * @param dataNode - Complete attestation data node
+ * @param globalRevocationLink - Hex string referencing revocation of the whole attestation
+ * @param privKey - Private key of signer
+ */
+export const getSignedDataNode = (
+  dataNode: IAttestation,
+  globalRevocationLink: string,
+  privKey: Buffer
+): IDataNode => {
+  const attestationNode: IAttestationNode = {
+    data: dataNode.data,
+    type: dataNode.type,
+    aux: dataNode.aux,
+    link: {
+      local: generateNonce(),
```

Comment:
Will this `local` link nonce need to be stored locally / in our db or anything?

Reply:
Everything returned by the getComponents function needs to be stored